docs/repoclosure.1
.\" repoclosure
.TH "repoclosure" "1" "13 January 2013" "" ""
.SH "NAME"
repoclosure \- display a list of unresolved dependencies for a yum repository
.SH "SYNOPSIS"
\fBrepoclosure\fP [options]
.SH "DESCRIPTION"
.PP
\fBrepoclosure\fP is a program that reads package metadata from one or more yum
repositories, checks all dependencies, and displays a list of packages with
unresolved dependencies.
.PP
.SH "OPTIONS"
.IP "\fB\-h, \-\-help\fP"
Display a help message, and then quit.
.IP "\fB\-c CONFIG, \-\-config=CONFIG\fP"
Config file to use (defaults to /etc/yum.conf).
.IP "\fB\-a ARCH, \-\-arch=ARCH\fP"
Check packages of the given archs, can be specified multiple times (default:
current arch).
.IP "\fB\-\-basearch=BASEARCH\fP"
Set the basearch for yum to run as.
.IP "\fB\-b, \-\-builddeps\fP"
Check build dependencies only (needs source repos enabled).
.IP "\fB\-l LOOKASIDE, \-\-lookaside=LOOKASIDE\fP"
Specify a lookaside repo id to query, can be specified multiple times.
.IP "\fB\-r REPOID, \-\-repoid=REPOID\fP"
Specify repo ids to query, can be specified multiple times (default is
all enabled).
.IP "\fB\-t, \-\-tempcache\fP"
Use a temp dir for storing/accessing yum-cache.
.IP "\fB\-q, \-\-quiet\fP"
Run quietly: no warnings printed to stderr.
.IP "\fB\-n, \-\-newest\fP"
Check only the newest packages in the repos.
.IP "\fB\-\-repofrompath=REPOID,PATH/URL\fP"
Specify a path or url to a repository (same path as in a baseurl) to add to
the repositories for this query. This option can be used multiple times. If
you want to view only the pkgs from this repository combine this with
\-\-repoid. The repoid for the repository is specified by REPOID.
.IP "\fB\-p PKG, \-\-pkg=PKG\fP"
Check closure for this package only
.IP "\fB\-g GROUP, \-\-group=GROUP\fP"
Check closure for packages in this group only

.PP
.SH "SEE ALSO"
.nf
.I yum.conf (5)
http://yum.baseurl.org/
.fi

.PP
.SH "AUTHORS"
.nf
See the Authors file included with this program.
.fi


ChangeLog
2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-03  Shawn Starr <shawn.starr@rogers.com>

	* docs/yum-utils.1: Update yum-utils man page, to add new utils.

2011-07-25  James Antill <james@and.org>

	Merge branch 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils  * 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils: (3 commits)
	typo leaving .conf in ybc.conf.yumvar should just be ybc.yumvar
	...

2011-07-25  James Antill <james@and.org>

	* plugins/aliases/aliases: Add some new aliases, fix chkD.

2011-07-25  Seth Vidal <skvidal@fedoraproject.org>

	* yum-config-manager.py: typo leaving .conf in ybc.conf.yumvar
	should just be ybc.yumvar

2011-07-12  James Antill <james@and.org>

	* plugins/security/security.py: Also auto-apply --security when
	given --sec-severity for other commands.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add repoquery @group completion.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* repoquery.py: Fix "nitmes" typo in previous commit.

2011-07-06  James Antill <james@and.org>

	* repoquery.py: Fix typo

2011-06-29  James Antill <james@and.org>

	* repoquery.py: Allow @foo groups as a packages selector in
	repoquery.

2011-06-25  Ville Skyttä <ville.skytta@iki.fi>

	* plugins/aliases/aliases.py, plugins/security/security.py,
	plugins/versionlock/versionlock.py: Remove some unused local
	variables and imports flagged by pyflakes.

2011-06-24  Ville Skyttä <ville.skytta@iki.fi>

	* debuginfo-install.py, package-cleanup.py, yum-builddep.py,
	yum-complete-transaction.py: Check doUtilBuildTransaction return
	value, exit early on failure.  BZ 716267

2011-06-22  James Antill <james@and.org>

	* plugins/security/security.py: Try to be clever when sorting the
	security severities.

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: After (external) merge is
	complete, give user option to delete other file

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: Support meld as well as vimdiff
	for merging  When using meld, the left side has the "other file"
	i.e. the file with the .rpmnew or .rpmsave extension. The right side
	has the "final file" i.e. the target file into which changes have to
	be merged.  If the rpm action was to save the existing conf file,
	then the left side will be the original conf file (.rpmsave) and the
	right side will be the new rpm-based configuration. The user should
	merge any changes they want to keep from their old .rpmsave from
	left to right.  If the rpm action was to keep the existing conf
	file, then the left side will be the new conf file (.rpmnew) and
	right side will be the current local configuration. The user should
	merge any changes in the new rpm file into their local config from
	left to right.  Ideally these operations would be 3-way diffs
	between the local file and the new rpm config, using the original
	rpm config as the common merge base. However, the original rpm
	config is not available to yum and therefore a 3-way diff is not
	possible. Perhaps a future enhancement would be to use a .orig file,
	if present, as the merge base for meld.  When using vimdiff, the
	order has not changed. The first parameter to vimdiff is the "final:
	file i.e. the target file, and the second parameter is the "other"
	file i.e. the .rpmnew or .rpmsave file.

2011-06-10  James Antill <james@and.org>

	* repodiff.py: Huge speedup for large number of add/remove obsoletes
	processing.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* yumdownloader.py: Stops yumdownloader from attempting to download
	the same package twice if it is available in multiple repos. BZ
	711767

2011-06-10  James Antill <james@and.org>

	* docs/repodiff.1, repodiff.py: Add --compare-arch for the other
	side of the multilib. diff. BZ 710579.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* repodiff.py: Changed archlist set up to handle multilib. BZ 710579


2011-06-01  James Antill <james@and.org>

	* docs/repoquery.1: Add releasever and setopt docs. from yum to
	repoquery. BZ 638542.

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: frelling pyyaml pkg named with weird case

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/puppetverify/puppetverify.conf,
	plugins/puppetverify/puppetverify.py, yum-utils.spec: add
	puppetverify plugin

2011-05-18  Frank Weltersbach <frankwe@live.com>

	* docs/package-cleanup.1: Fix typo in package-cleanup man page. BZ
	705617.

2011-05-17  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py:     - fix rh bug: #705179     - fix fmt_queryformat not being happy
	    with new 'output' formats passed       everywhere else.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Allow files/provides to be used in
	yumdownloader.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Exit with code 2, if we can't download the
	packages asked for. BZ 699469.

2011-04-26  James Antill <james@and.org>

	* plugins/auto-update-debuginfo/auto-update-debuginfo.py: try/except
	the open, for the rpmdbv cached data (a-u-d plugin).

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add yum-debug-dump completion.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Mention filename arg in yum-debug-dump usage
	message.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Make yum-debug-dump.py executable for running
	in-place.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Use functions from recent bash-completion if
	available.

2011-04-06  James Antill <james@and.org>

	* repoquery.py, yum-groups-manager.py: Convert

repoclosure.py
#!/usr/bin/python -t

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
# seth vidal 2005 (c) etc etc


#Read in the metadata of a series of repositories and check all the
#   dependencies in all packages for resolution. Print out the list of
#   packages with unresolved dependencies

import sys
import os

import logging
import yum
import yum.Errors
from yum.misc import getCacheDir
from optparse import OptionParser
import rpmUtils.arch
import rpmUtils.updates
from yum.constants import *
from yum.packageSack import ListPackageSack

def parseArgs():
    usage = """
    Read in the metadata of a series of repositories and check all the   
    dependencies in all packages for resolution. Print out the list of
    packages with unresolved dependencies
    
    %s [-c <config file>] [-a <arch>] [-l <lookaside>] [-r <repoid>] [-r <repoid2>]
    """ % sys.argv[0]
    parser = OptionParser(usage=usage)
    parser.add_option("-c", "--config", default='/etc/yum.conf',
        help='config file to use (defaults to /etc/yum.conf)')
    parser.add_option("-a", "--arch", default=[], action='append',
        help='check packages of the given archs, can be specified multiple ' +
             'times (default: current arch)')
    parser.add_option("--basearch", default=None, 
                      help="set the basearch for yum to run as")
    parser.add_option("-b", "--builddeps", default=False, action="store_true",
        help='check build dependencies only (needs source repos enabled)')
    parser.add_option("-l", "--lookaside", default=[], action='append',
        help="specify a lookaside repo id to query, can be specified multiple times")
    parser.add_option("-r", "--repoid", default=[], action='append',
        help="specify repo ids to query, can be specified multiple times (default is all enabled)")
    parser.add_option("-t", "--tempcache", default=False, action="store_true", 
        help="Use a temp dir for storing/accessing yum-cache")
    parser.add_option("-q", "--quiet", default=0, action="store_true", 
                      help="quiet (no output to stderr)")
    parser.add_option("-n", "--newest", default=0, action="store_true",
                      help="check only the newest packages in the repos")
    parser.add_option("--repofrompath", action="append",
                      help="specify repoid & paths of additional repositories - unique repoid and path required, can be specified multiple times. Example. --repofrompath=myrepo,/path/to/repo")
    parser.add_option("-p", "--pkg", action="append",
                      help="check closure for this package only")
    parser.add_option("-g", "--group", action="append",
                      help="check closure for packages in this group only")
    (opts, args) = parser.parse_args()
    return (opts, args)

#  Note that this is a "real" API, used by spam-o-matic etc.
# so we have to do at least some API guarantee stuff.
class RepoClosure(yum.YumBase):
    def __init__(self, arch=[], config="/etc/yum.conf", builddeps=False, pkgonly=None,
                 basearch=None, grouponly=None):
        yum.YumBase.__init__(self)
        if basearch:
            self.preconf.arch = basearch
        self.logger = logging.getLogger("yum.verbose.repoclosure")
        self.lookaside = []
        self.builddeps = builddeps
        self.pkgonly = pkgonly
        self.grouponly = grouponly
        self.doConfigSetup(fn = config,init_plugins=False)
        self._rc_arches = arch

        if hasattr(self.repos, 'sqlite'):
            self.repos.sqlite = False
            self.repos._selectSackType()
    
    def evrTupletoVer(self,tup):
        """convert an evr tuple to a version string, return None if nothing
        to convert"""
    
        e, v, r = tup

        if v is None:
            return None
    
        val = v
        if e is not None:
            val = '%s:%s' % (e, v)
    
        if r is not None:
            val = '%s-%s' % (val, r)
    
        return val
    
    def readMetadata(self):
        self.doRepoSetup()
        archs = []
        if not self._rc_arches:
            archs.extend(self.arch.archlist)
        else:
            for arch in self._rc_arches:
                archs.extend(self.arch.get_arch_list(arch))

        if self.builddeps and 'src' not in archs:
            archs.append('src')
        self.doSackSetup(archs)
        for repo in self.repos.listEnabled():
            self.repos.populateSack(which=[repo.id], mdtype='filelists')

    def getBrokenDeps(self, newest=False):
        unresolved = {}
        resolved = {}
        pkgs = self.pkgSack
        if newest:
            pkgs = self.pkgSack.returnNewestByNameArch()
            mypkgSack = ListPackageSack(pkgs)
            pkgtuplist = mypkgSack.simplePkgList()
            
            # toss out any of the obsoleted pkgs so we can't depsolve with them
            self.up = rpmUtils.updates.Updates([], pkgtuplist)
            self.up.rawobsoletes = mypkgSack.returnObsoletes()
            for pkg in pkgs:
                fo = self.up.checkForObsolete([pkg.pkgtup])
                if fo:
                    # useful debug to make sure the obsoletes is sane
                    #print "ignoring obsolete pkg %s" % pkg
                    #for i in fo[pkg.pkgtup]:
                    #    print i
                    self.pkgSack.delPackage(pkg)

            # we've deleted items so remake the pkgs
            pkgs = self.pkgSack.returnNewestByNameArch()
            pkgtuplist = mypkgSack.simplePkgList()
        
        if self.builddeps:
            pkgs = filter(lambda x: x.arch == 'src', pkgs)

        pkglist = self.pkgonly
        if self.grouponly:
            if not pkglist:
                pkglist = []
            for group in self.grouponly:
                groupobj = self.comps.return_group(group)
                if not groupobj:
                    continue
                pkglist.extend(groupobj.packages)

        if pkglist:
            pkgs = filter(lambda x: x.name in pkglist, pkgs)

        for pkg in pkgs:
            if pkg.repoid in self.lookaside:
                # don't attempt to resolve dependency issues for
                # packages from lookaside repositories
                continue
            for (req, flags, (reqe, reqv, reqr)) in pkg.returnPrco('requires'):
                if req.startswith('rpmlib'): continue # ignore rpmlib deps
            
                ver = self.evrTupletoVer((reqe, reqv, reqr))
                if (req,flags,ver) in resolved:
                    continue
                
                resolve_sack = [] # make it empty
                try:
                    resolve_sack = self.whatProvides(req, flags, ver)
                except yum.Errors.RepoError, e:
                    pass
            
                if len(resolve_sack) < 1:
                    if pkg not in unresolved:
                        unresolved[pkg] = []
                    unresolved[pkg].append((req, flags, ver))
                    continue
                    
                if newest:
                    resolved_by_newest = False
                    for po in resolve_sack:# look through and make sure all our answers are newest-only
                        if po.pkgtup in pkgtuplist:
                            resolved_by_newest = True
                            break

                    if resolved_by_newest:                    
                        resolved[(req,flags,ver)] = 1
                    else:
                        if pkg not in unresolved:
                            unresolved[pkg] = []
                        unresolved[pkg].append((req, flags, ver))
                        
        return unresolved
    

def main():
    (opts, cruft) = parseArgs()
    my = RepoClosure(arch=opts.arch, 
                     config=opts.config, 
                     builddeps=opts.builddeps,
                     pkgonly=opts.pkg,
                     grouponly=opts.group,
                     basearch=opts.basearch)

    if opts.repofrompath:
        # setup the fake repos
        for repo in opts.repofrompath:
            repoid,repopath = tuple(repo.split(','))
            if repopath.startswith('http') or repopath.startswith('ftp') or repopath.startswith('file:'):
                baseurl = repopath
            else:
                repopath = os.path.abspath(repopath)                
                baseurl = 'file://' + repopath
                
            newrepo = yum.yumRepo.YumRepository(repoid)
            newrepo.name = repopath
            newrepo.baseurl = baseurl
            newrepo.basecachedir = my.conf.cachedir
            newrepo.metadata_expire = 0
            newrepo.timestamp_check = False
            my.repos.add(newrepo)
            my.repos.enableRepo(newrepo.id)
            my.logger.info( "Added %s repo from %s" % (repoid,repopath))
    
    if opts.repoid:
        for repo in my.repos.repos.values():
            if ((repo.id not in opts.repoid) and
                (repo.id not in opts.lookaside)):
                repo.disable()
            else:
                repo.enable()

    if opts.lookaside:
        my.lookaside = opts.lookaside

    if os.geteuid() != 0 or opts.tempcache:
        cachedir = getCacheDir()
        if cachedir is None:
            my.logger.error("Error: Could not make cachedir, exiting")
            sys.exit(50)
            
        my.repos.setCacheDir(cachedir)

    if not opts.quiet:
        my.logger.info('Reading in repository metadata - please wait....')

    try:
        my.readMetadata()
    except yum.Errors.RepoError, e:
        my.logger.info(e)
        my.logger.info('Some dependencies may not be complete for this repository')
        my.logger.info('Run as root to get all dependencies or use -t to enable a user temp cache')

    if not opts.quiet:
        my.logger.info('Checking Dependencies')

    baddeps = my.getBrokenDeps(opts.newest)
    if opts.newest:
        num = len(my.pkgSack.returnNewestByNameArch())
    else:
        num = len(my.pkgSack)
        
    repos = my.repos.listEnabled()

    if not opts.quiet:
        my.logger.info('Repos looked at: %s' % len(repos))
        for repo in repos:
            my.logger.info('   %s' % repo)
        my.logger.info('Num Packages in Repos: %s' % num)
    
    pkgs = baddeps.keys()
    
    def sortbyname(a,b):
        return cmp(a.__str__(),b.__str__())
    
    pkgs.sort(sortbyname)
    
    for pkg in pkgs:
        my.logger.info('package: %s from %s\n  unresolved deps: ' % (pkg, pkg.repoid))
        for (n, f, v) in baddeps[pkg]:
            req = '%s' % n
            if f: 
                flag = LETTERFLAGS[f]
                req = '%s %s'% (req, flag)
            if v:
                req = '%s %s' % (req, v)
            
            my.logger.info('     %s' % req)
    if baddeps:
        sys.exit(1)

if __name__ == "__main__":
    try:
        main()
    except (yum.Errors.YumBaseError, ValueError), e:
        print >> sys.stderr, str(e)
        sys.exit(2)
